### PR TITLE
Add update_token_pool_internal entrypoint spec

### PR DESCRIPTION
**Pull request by sskeirik**
Note: this proof is still broken due to the negative case proofs not going through. As far as I can tell, these proofs get stuck because the backend is not able to prune unsatisfiable states, but I may have not tightened some initial condition enough.

EDIT: I would appreciate others looking into the negative case proofs and seeing if the final states do actually seem to have unsatisfiable path conditions.

Fixes: #220

### CODE REVIEW

**Comment by hjorthjort** [comment]
I'll put off review on this until we have #234 merged, so that we can make this PR smaller

**Comment by ehildenb** [comment]
- Double check that updating to master doesn't fix this.
- Double check local build is using submodule.
- Doube check z3 versions.
- Try `archiveArtifacts` on teh `kore-exec.tar.gz` and download it from the Artifacts page on CI. Check K Jenkinsfile for how to do (it does the same with kserver.log).

File: tests/proofs/dexter/dexter-spec.md [hunk 1]
```diff
@@ -267,3 +267,119 @@ If any of the conditions are not satisfied, the call fails.
 ```k
 endmodule
 ```
+
+## Update Token Pool Internal
+
+```k
+module DEXTER-UPDATETOKENPOOLINTERNAL-SPEC
+  imports DEXTER-VERIFICATION
+```
```
Comment (ehildenb) [inline tests/proofs/dexter/dexter-spec.md:276]: Let's make sure we state that we need to be sure (on paper) that the list of claims here is exhaustive.

File: tests/proofs/dexter/dexter.md [hunk 2]
```diff
@@ -83,6 +84,12 @@ configuration <dexterTop>
 We define our list of entrypoints below.
 Each entrypoint is given a unique abstract parameter type that we use to simplify our proof structure.
 
+Summary: The underlying token contract updates the Dexter contract's view of its own token balance if the following conditions are satisifed:
+
+         1.  the token pool _is_ currently updating (i.e. `storage.selfIsUpdatingTokenPool = true`)
+         2.  exactly 0 tez was transferred to this contract when it was invoked
+         3.  if using version FA2, the input parameter list is non-empty.
+
```
Comment (hjorthjort) [inline tests/proofs/dexter/dexter.md:92]:
```suggestion
```

Looks like this got pasted in the wrong place. It's already pasted in the right place (in the spec file).

File: tests/proofs/lemmas.md [hunk 3]
```diff
@@ -13,7 +13,7 @@ requires "michelson.md"
 ```k
 module LEMMAS
   imports MICHELSON
-
+
```
Comment (hjorthjort) [inline tests/proofs/lemmas.md:16]:
```suggestion

```

Accidental whitespace

File: tests/proofs/dexter/dexter.md [hunk 4]
```diff
@@ -16,6 +16,7 @@ We begin start our verification project by opening a new module context in which
 
 ```k
```
Comment (hjorthjort) [inline tests/proofs/dexter/dexter.md:17]: Accidental whitespace?

File: tests/proofs/dexter/dexter.md [hunk 5]
```diff
@@ -16,6 +16,7 @@ We begin start our verification project by opening a new module context in which
...
 requires "../lemmas.md"
+
```
Comment (hjorthjort) [inline tests/proofs/dexter/dexter.md:19]: Accidental whitespace?

**Comment by nishantjr** [comment]
Ready to review.

**Comment by hjorthjort** [comment]
Looks like you have a Makefile issue:
```

[Error] Critical: Module DEXTER-REMOVELIQUIDITY-SPEC does not exist.
```

File: Makefile [hunk 6]
```diff
@@ -382,15 +382,17 @@ tests/%.prove: tests/% $(prove_kompiled)
...
 dexter_spec_modules = DEXTER-SPEC                          \
-                      DEXTER-ADDLIQUIDITY-POSITIVE-SPEC    \
                       DEXTER-ADDLIQUIDITY-NEGATIVE-SPEC    \
-                      DEXTER-REMOVELIQUIDITY-POSITIVE-SPEC \
+                      DEXTER-ADDLIQUIDITY-POSITIVE-SPEC    \
+                      DEXTER-DEFAULT-SPEC                  \
                       DEXTER-REMOVELIQUIDITY-NEGATIVE-SPEC \
-                      DEXTER-SETMANAGER-SPEC               \
+                      DEXTER-REMOVELIQUIDITY-POSITIVE-SPEC \
+                      DEXTER-REMOVELIQUIDITY-SPEC          \
```
Comment (hjorthjort) [inline Makefile:390]:
```suggestion
```

Comment (nishantjr) [reply Makefile]: Thanks

File: tests/proofs/dexter/dexter.md [hunk 7]
```diff
@@ -669,6 +663,7 @@ If all steps are completed, only the Dexter-specific storage is updated.
 ## Miscellaneous Lemmas
 
 ```k
+    rule notBool(notBool(BExp)) => BExp [simplification]
```
Comment (hjorthjort) [inline tests/proofs/dexter/dexter.md:666]:
```suggestion
    rule notBool notBool BExp => BExp [simplification, smt-lemma]
```

Comment (hjorthjort) [inline tests/proofs/dexter/dexter.md:666]:
```suggestion
    rule notBool notBool BExp => BExp [simplification]
    #Not ( #Not ( X ) ) => X [anywhere, simplification]
```

File: tests/proofs/dexter/dexter-spec.md [hunk 8]
```diff
@@ -469,6 +470,8 @@ NOTE: The failure conditions are split into two claims with identical configurat
      orBool Sender =/=K Source
 ```
 
+TODO: This seems redundant
+
```
Comment (hjorthjort) [inline tests/proofs/dexter/dexter-spec.md:474]:
```suggestion
```

File: tests/proofs/dexter/dexter-spec.md [hunk 9]
```diff
@@ -267,3 +267,145 @@ If any of the conditions are not satisfied, the call fails.
...
+        <selfIsUpdatingTokenPool> true => false </selfIsUpdatingTokenPool>
+        <myamount> #Mutez(0) </myamount>
+        <tokenPool> _ => Balance </tokenPool>
+        <tokenAddress> TokenAddress </tokenAddress>
+        <senderaddr> TokenAddress </senderaddr>
+```
+
+```k
+  claim <k> #runProof(true, UpdateTokenPoolInternalFA2([ Pair Pair _ _ Balance ] ;; _BalanceOfResultRest)) => . </k>
+        <stack> .Stack </stack>
```
[the remark this answers is not in the document]
Comment (hjorthjort) [reply tests/proofs/dexter/dexter-spec.md:548]: Answer: the balance request that is sent is a singleton list, and here the first result back is read.

**Review by hjorthjort — APPROVED** [review]

File: Jenkinsfile [hunk 10]
```diff
@@ -19,12 +19,17 @@ pipeline {
...
         stage('Unit')             { steps { sh 'make test-unit     -j8' } }
         stage('Symbolic')         { steps { sh 'make test-symbolic -j2' } }
         stage('Prove')            { steps { sh 'make test-prove    -j2' } }
-        stage('Dexter Proofs')    { steps { sh 'make dexter-prove  -j8' } }
+      }
+    }
+    stage('Integration Proofs') { 
+      options { timeout(time: 180, unit: 'MINUTES') }
+      parallel {
+        stage('Dexter Proofs') { steps { sh 'make dexter-prove  -j8' } }
```
Comment (hjorthjort) [inline Jenkinsfile:32]:
```suggestion
        stage('Dexter Proofs') { steps { sh 'make dexter-prove  -j4' } }
```

**Pull request merged**